MM-60441: Re-index public channels when a user joins a team

agarciamontoro opened the pull request:
#### Summary

Before this PR, when a user joined a team, we didn't update the channels indexed in Elasticsearch with the new membership, which caused the new channels not to show up in the channel browser when the user looked for them (at least not immediately, eventually they would show up due to other index events).

This PR, based off @isacikgoz's work in #29165, fixes that by indexing all public channels in bulk when we add a user to a team in `SearchTeamStore.SaveMember`. This PR has been in the works for too long (see [this document summarizing everything](https://mattermost.atlassian.net/wiki/spaces/XYZ/pages/3706585097/MM-60441+New+users+cannot+search+channels)), so I've created a new branch consolidating all the work we did in multiple places. The commits present at the time of opening the PR are:

- **Original work** - This is Ibrahim's work. I didn't want his effort to be lost, so I squashed all of his commits here:
    - e9c1e67: Index all public channels when a user joins a team
- **Original fixes** - The original work had some bugs, so the next commits fix those (see an analysis in [the Previous work section of the document](https://mattermost.atlassian.net/wiki/spaces/XYZ/pages/3706585097/MM-60441+New+users+cannot+search+channels#Previous-work))
    - 9fe77ed: Precompute team members for indexChannelsForTeam
    - f9a7282: Refactor RequestContextWithMaster to store package
    - c17d243: Use master in call to GetTeamMembersForChannel
    - 1a16c82: Fix GetPublicChannelsForTeam paging
- **New work** - The final fix consisted in a bulk-index of all public channels, instead of iterating through all of them and indexing them individually. See [the Analysis section of the document](https://mattermost.atlassian.net/wiki/spaces/XYZ/pages/3706585097/MM-60441+New+users+cannot+search+channels#Analysis). The next commits are focused on that work, adding the necessary infrastructure (the sync bulk indexer), implementing the `SyncBulkIndexChannels` function (for both Elasticsearch and Opensearch) and testing all the work (which was assisted by AI). Finally, we actually use all this work by using the new `SyncBulkIndexChannels` function from `indexChannelsForTeam`:
    - bfe4671: Add a synchronous bulk indexer for Opensearch
    - 0d9fc6b: Implement Opensearch's SyncBulkIndexChannels
    - 6643ae3: Add a synchronous bulk indexer for Elasticsearch
    - 8bc121e: Implement Elasticsearch's SynkBulkIndexChannels
    - f81207f: Test SyncBulkIndexChannels
    - fc7c182: make mocks
    - eac5a16: Bulk index channels on indexChannelsForTeam

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-60441

#### Screenshots
--

#### Release Note
```release-note
Fixed a bug that prevented new users from searching channels right after joining a team when Elasticsearch was enabled.
```

agarciamontoro commented:
@DHaussermann, I've added you here early so that this PR is in your radar. I'll add some QA test steps in the ticket.

streamer45 commented:
Looking forward to seeing the `Mana Spent` field value for this ticket.

agnivade commented:
One good thing is that the ES/OS code is now open source, so it definitely helps with reviewing the code. :D

agnivade commented on server/enterprise/elasticsearch/elasticsearch/bulk.go:
```diff
@@ -69,12 +79,20 @@ func (r *Bulk) DeleteOp(op types.DeleteOperation) error {
 // It MUST be called with an already acquired mutex.
 func (r *Bulk) flushIfNecessary() error {
+	// Flush always if in sync mode
+	if r.sync {
```
Do we still need this flag? It is already checked at the higher level.

agarciamontoro replied:
We can remove it, it was some defensive code but it's true that it's not necessary.

agarciamontoro replied:
Fixed in https://github.com/mattermost/mattermost/pull/33400/commits/126e60c6e32c95c94f78986a8544576d3e537445.

agnivade commented on server/enterprise/elasticsearch/opensearch/bulk.go:
```diff
@@ -126,6 +144,10 @@ func (r *Bulk) Stop() error {
 }
 
 func (r *Bulk) periodicFlusher() {
+	if r.sync {
```
Same comment as earlier.

agnivade commented on server/enterprise/elasticsearch/elasticsearch/elasticsearch.go:
```diff
@@ -748,6 +754,49 @@ func (es *ElasticsearchInterfaceImpl) IndexChannel(rctx request.CTX, channel *mo
+		}
+	}
+
+	if err := es.syncBulkProcessor.flush(); err != nil {
```
I guess the only slight problem with the syncProcessor is we are not flushing at all until the very end, which means that potentially a lot of items will be in memory until they are flushed. There's also another problem of very large request sizes on which there is a limit in ES.

Wondering if we can keep the periodicFlusher intact (which will ensure flushing when batch size exceeds a threshold), but also add a force flush at the end which should take care of any pending items in the buffer?

agnivade replied:
> Maybe a simple solution would be to add another condition to the timer that flushes the data every, let's say, 10MiB. Thoughts?

That could certainly be a solution and probably what the old bulk indexer had. But even with the current solution, if `LiveIndexingBatchSize` is too high, we can always ask the customer to lower it which would make it flush earlier. Alternatively, I think tracking `r.buf.Len()` to check the size of the buffer and then flush it when it crosses 10MiB is also fine. It's not that much code.

agarciamontoro commented:
Somewhat big refactor in the latest four commits: I've reverted the changes adding the synchronous bulk indexers in both Elasticsearch and Opensearch, and I've cleaned the code so that the caller only has to decide what threshold they're going to use for the bulk client to know when to flush: size of the request, number of requests and a simple timer.

The problem with these three thresholds happens on Elasticsearch: The two ways of managing bulk indexers in Elasticsearch are:
1. Using the high-level `go-elasticsearch/v8/esutil.BulkIndexer` type.
2. Using the lower-level `go-elasticsearch/v8/typedapi/code/bulk.Bulk` type.

The first one supports flushing when the size of the request exceeds a threshold, and by time, but not by number of requests. The second one is more basic, and it simply exposes and API to add operations and flush manually (the `Do` function), so we can implement number of requests and timer thresholds on top. The problem with the second one is that it doesn't expose the underlying buffer, so we cannot check its size.

That is why I have decided to implement two `BulkClient`s for Elasticsearch, both created via the `elasticsearch.NewBulk` function, that hides all this complexity and simply lets the caller decide which threshold to use.

The implementation for Opensearch is more custom, so all three thresholds can be accommodated in the same implementation.

The writing of the tests was assisted by Claude.

agnivade commented on server/enterprise/elasticsearch/elasticsearch/bulk_client_data.go:
```diff
@@ -0,0 +1,217 @@
+		OnFlushEnd: func(context.Context) {
+			logger.Debug("elasticsearch bulk indexer flush ended")
+		},
+		// Index:         "",
+		// ErrorTrace:    false,
+		// FilterPath:    []string{},
+		// Header: map[string][]string{
+		// 	"": {},
+		// },
+		// Human:               false,
+		// Pipeline:            "",
+		// Pretty:              false,
+		// Refresh:             "",
+		// Routing:             "",
+		// RequireAlias:        false,
+		// Source:              []string{},
+		// SourceExcludes:      []string{},
+		// SourceIncludes:      []string{},
+		// Timeout:             0,
+		// WaitForActiveShards: "",
```
Remove these?

agnivade commented on server/enterprise/elasticsearch/elasticsearch/bulk_client_data.go:
```diff
@@ -0,0 +1,217 @@
+		Index:      *op.Index_,
+		Action:     "index",
+		DocumentID: *op.Id_,
+		// Routing:         "",
```
Looks like more debug statements.

agnivade commented on server/enterprise/elasticsearch/elasticsearch/bulk_client_data.go:
```diff
@@ -0,0 +1,217 @@
+	b.mut.Lock()
+	defer b.mut.Unlock()
+
+	b.logger.Info("Stopping Bulk processor")
+
+	ctx, cancel := context.WithTimeout(context.Background(), b.reqTimeout)
+	defer cancel()
+
+	return b.indexer.Close(ctx)
```
Could we refactor this part from here and Flush into _stop?

agnivade reviewed (approved):
Thanks!

streamer45 reviewed (approved):
Nice work! I just left a few minor comments and suggestions. Feel free to take them or leave them :)

agarciamontoro commented on server/channels/store/searchlayer/layer.go:
```diff
@@ -111,6 +112,35 @@ func (s *SearchStore) indexUser(rctx request.CTX, user *model.User) {
+
+	// Use master context to avoid replica lag issues when reading team members
+	masterRctx := store.RequestContextWithMaster(rctx)
+	teamMemberIDs, err := s.channel.GetTeamMembersForChannel(masterRctx, channels[0].Id)
```
The SQL query uses the team ID from that channel. All members of `channels` are in the same team, so we use the first one, which we know exists because we've checked right before that.

agarciamontoro commented on server/channels/store/searchlayer/team_layer.go:
```diff
@@ -17,7 +17,11 @@ type SearchTeamStore struct {
 	if err == nil {
-		s.rootStore.indexUserFromID(rctx, member.UserId)
+		// Nothing to do if search engine is not active
+		if s.rootStore.searchEngine.ActiveEngine() != "database" && s.rootStore.searchEngine.ActiveEngine() != "none" {
```
Doing this in another PR.

agarciamontoro commented on server/enterprise/elasticsearch/elasticsearch/bulk.go:
```diff
@@ -4,130 +4,58 @@
+}
+
+// NewBulk returns a BulkClient, with the specific implementation depending on
+// the specified thersholds in bulkSettings.
```
Ah, yes, theirs holds.

agarciamontoro commented on server/enterprise/elasticsearch/elasticsearch/bulk.go:
```diff
@@ -4,130 +4,58 @@
-	if r.pendingRequests > *r.settings.LiveIndexingBatchSize {
-		return r._flush()
+) (BulkClient, error) {
+	if bulkSettings.FlushBytes == 0 &&
```
If we ever use this code elsewhere, we can refactor, but I'm ok keeping it there.

agarciamontoro commented on server/enterprise/elasticsearch/elasticsearch/bulk.go:
```diff
@@ -4,130 +4,58 @@
-		return r._flush()
+) (BulkClient, error) {
+	if bulkSettings.FlushBytes == 0 &&
+		bulkSettings.FlushInterval == 0 &&
+		bulkSettings.FlushNumReqs == 0 {
```
We could, but I think being explicit makes it easier for a human to parse it.

agarciamontoro commented on server/enterprise/elasticsearch/elasticsearch/bulk_client_data.go:
```diff
@@ -0,0 +1,217 @@
+
+	var bodyReader io.ReadSeeker
+	switch v := doc.(type) {
+	case []byte:
+		bodyReader = bytes.NewReader(v)
+	case json.RawMessage:
+		bodyReader = bytes.NewReader(v)
```
Ha! I thought the same! But in that case `v` is bind to an `any` interface, and `NewReader` expects a `bytes.Reader`. We could cast it, but then the benefit of less code is gone.

agarciamontoro commented on server/enterprise/elasticsearch/elasticsearch/sync_bulk.go:
```diff
@@ -0,0 +1,94 @@
+func (r *SyncBulk) IndexOp(op types.IndexOperation, doc any) error {
+	var body io.ReadSeeker
+	switch v := doc.(type) {
+	case []byte:
+		body = bytes.NewReader(v)
+	case json.RawMessage:
+		body = bytes.NewReader(v)
```
Same as before

agarciamontoro commented on server/enterprise/elasticsearch/elasticsearch/bulk_client_data.go:
```diff
@@ -0,0 +1,217 @@
+	case json.RawMessage:
+		bodyReader = bytes.NewReader(v)
+	default:
+		body, err := json.Marshal(doc)
+		if err != nil {
+			return err
+		}
+		bodyReader = bytes.NewReader(body)
```
I considered it, but I don't think this is needed for now

agarciamontoro commented on server/enterprise/elasticsearch/elasticsearch/sync_bulk.go:
```diff
@@ -0,0 +1,94 @@
+		body = bytes.NewReader(data)
+	}
+
+	return r.bulkIndexer.Add(context.Background(), esutil.BulkIndexerItem{
```
There is a [default timeout of 1 minute](https://www.elastic.co/guide/en/elasticsearch/reference/8.18/docs-bulk.html).

agarciamontoro commented on server/enterprise/elasticsearch/opensearch/bulk.go:
```diff
@@ -101,10 +107,20 @@ func (r *Bulk) DeleteOp(op *types.DeleteOperation) error {
 // It MUST be called with an already acquired mutex.
 func (r *Bulk) flushIfNecessary() error {
+	// Check data threshold, only if specified
+	if r.bulkSettings.FlushBytes > 0 {
+		if r.buf.Len() >= r.bulkSettings.FlushBytes {
```
They are different kinds of checks, right? The first one is "is this feature enabled"? The second one is: have we crossed the threshold? It's of course the same, but in my head this looks better :shrug:

agarciamontoro commented on server/enterprise/elasticsearch/opensearch/bulk.go:
```diff
@@ -101,10 +107,20 @@ func (r *Bulk) DeleteOp(op *types.DeleteOperation) error {
-	if r.pendingRequests > *r.settings.LiveIndexingBatchSize {
-		return r._flush()
+	// Check number of requests threshold, only if specified
+	if r.bulkSettings.FlushNumReqs > 0 {
+		if r.pendingRequests > r.bulkSettings.FlushNumReqs {
```
Same

agarciamontoro commented:
@DHaussermann, this is ready for QA review.

DHaussermann commented:
I have now done a round of testing.
- Added user to team and ensured channels are findable. 
- This includes channels created within a within a minute or so of the new user joining the team
- Tested with user joining via invite code as well as a bulk team invite from an existing user
- Tested with Opensearch and Elastic Search   

@agarciamontoro this looks good. Just a couple follow up questions:

1. Can we be sure we're not unnecessary re-indexing when a bulk invite happens? As per comment above it's not quite clear to me if this is occurring.
2. Do I need to test this with a more significant data set containing thousands of channels to see if there are performance issues?

agarciamontoro commented:
Thank you for testing this! Some comments:

>  I have also tried inviting sevral users to a team via the UI to ensure the indexing happens only once when you bulk invite but it's hard to tell as indexing requests keep going for a bit after the invite. Can you please clarify the behavior when you bulk invite?

When you bulk invite many users, the indexing will happen once per user. The bulk indexing I mention in the PR description and comments refers to the indexing of all the public channels of the team, which are indexed in bulk. But that happens for every user that joins a team. This is of course not optimal, but the solution to index just once when adding many users to a team at the same time is way more involved and it was decided early on to tackle it in a separate PR if needed.

> Also, do we need to audit somewhere that this indexing jobs have happened?

If the channels appear in the channel browser immediately, that's enough proof that the indexing happened. To be extra sure, you could take a look at Elasticsearch/Opensearch logs, but if all is good, I don't think that's necessary 

> 1. Can we be sure we're not unnecessary re-indexing when a bulk invite happens? As per comment above it's not quite clear to me if this is occurring.

Answered above. We do re-index when a bulk invite happens. This is a known short-coming of this PR, but will fixed in the future.

> 2. Do I need to test this with a more significant data set containing thousands of channels to see if there are performance issues?

Yes, please. I saw some bugs at the beginning that were only reproducible with a large dataset. If you can test it with the 100M-posts database, that'd be perfect. I can help you set it up next week (starting Tuesday, I'm out Monday).

agarciamontoro commented:
@matthewbirtch, it makes sense! I created the ticket, summarizing the current state and the possible solutions I can think of as of now: https://mattermost.atlassian.net/browse/MM-65132 Thanks for taking a look!

matthewbirtch commented:
> @matthewbirtch, it makes sense! I created the ticket, summarizing the current state and the possible solutions I can think of as of now: https://mattermost.atlassian.net/browse/MM-65132 Thanks for taking a look!

Thanks @agarciamontoro!

DHaussermann reviewed (approved):
Tested and passed as per testing above ☝️.
- This PR does resolve this issue
- The times until channels can be found has seen a massive improvement
- With very large data sets, there is a small remaining delay. We will consider a UX change in the future to address this. 

LGTM!

Thanks @agarciamontoro for providing the details of your investigation here!
